### PR TITLE
Add new view to edit dictionaries.

### PR DESCRIPTION
**Pull request by jonathanrocher**
Create a general class to edit dictionary values in a non-default way.

```
    from traits.api import Range

    def value_to_trait(value):
        return Range(low=0, high=1, value=value)

    editor = DictValueEditor(target_dict={"a": 1, "b c": 0.5},
                             value_to_trait=value_to_trait,
                             title="Dict editor", show_title_in_view=True)
    editor.configure_traits()
```
Screen Shot 2021-01-14 at 9 21 11 AM

### CODE REVIEW

File: etstool.py [hunk 1]
```diff
@@ -508,7 +508,7 @@ def get_parameters(runtime, toolkit, environment, **adtl_params):
         raise RuntimeError(msg.format(**parameters))
 
     if environment is None:
-        env_pattern = PKG_NAME + '-test-{toolkit}-py{runtime}'
+        env_pattern = PKG_NAME + '-develop-{toolkit}-py{runtime}'
```
Comment (jonathanrocher) [inline etstool.py:511]: Unrelated.

File: app_common/traitsui/dict_editors.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    def traits_view(self):
+        item_list = []
+        if self.title and self.show_title_in_view:
+            item_list.append(make_window_title_group(self.title))
+
+        item_list += [Item(trait_name, **self.item_kwargs(key))
+                      for trait_name, key in self._trait_map.items()]
+        view = self.view_klass(
+            *item_list,
+            buttons=OKCancelButtons,
```
Comment (chamshop) [inline app_common/traitsui/dict_editors.py:70]: We should only have a `Cancel` button if we are actually providing the functionality that cancels the actions taken in the view.

File: app_common/traitsui/dict_editors.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            self._key_map[key] = trait_name
+            self._trait_map[trait_name] = key
+            self.observe(self.update_target_dict, trait_name)
+
+    def traits_view(self):
+        item_list = []
+        if self.title and self.show_title_in_view:
+            item_list.append(make_window_title_group(self.title))
+
+        item_list += [Item(trait_name, **self.item_kwargs(key))
```
Comment (chamshop) [inline app_common/traitsui/dict_editors.py:66]: If possible, I'd like to see  you making use of this `item_kwargs` callable within the demo code at the bottom of the file.

File: app_common/traitsui/dict_editors.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            title=self.title
+        )
+        return view
+
+    def update_target_dict(self, event):
+        """ A trait was changed: update the dictionary.
+        """
+        self.target_dict[self._trait_map[event.name]] = event.new
+
+    def set_values(self, **kwargs):
```
Comment (chamshop) [inline app_common/traitsui/dict_editors.py:80]: I do not see where this function is actually used anywhere in our consuming code...

**Review by chamshop — APPROVED** [review]
Looks good, a few questions and comments though.
I would consider removing `set_values` if there is no current consumer of it.

File: app_common/traitsui/dict_editors.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    def __init__(self, *args, **traits):
+        if args:
+            if len(args) != 1:
+                msg = "A DictValueEditor should be created with 1 " \
+                      "positional argument at the most, providing the " \
+                      "target_dict."
+                raise ValueError(msg)
+            else:
+                traits["target_dict"] = args[0]
```
[the remark this answers is not in the document]
Comment (jonathanrocher) [reply app_common/traitsui/dict_editors.py:47]: No because then it's required to pass it as an argument. I want to support passing it as an arg or a kwarg.

File: app_common/traitsui/dict_editors.py (hunk 2 above)
Comment (jonathanrocher) [reply app_common/traitsui/dict_editors.py:70]: I didn't implement the possibility to cancel what was done. I guess I could.

File: app_common/traitsui/dict_editors.py (hunk 4 above)
Comment (jonathanrocher) [reply app_common/traitsui/dict_editors.py:80]: It is there, used by the button which resets all values to `0.5`.

File: app_common/traitsui/dict_editors.py (hunk 3 above)
Comment (jonathanrocher) [reply app_common/traitsui/dict_editors.py:66]: OK. Even better, I will add a test.

File: app_common/traitsui/dict_editors.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    key_to_trait = Callable
+
+    value_to_trait = Callable(Float)
+
+    item_kwargs = Callable(lambda key: {"label": key})
+
+    title = Str
+
+    show_title_in_view = Bool
```
[the remark this answers is not in the document]
Comment (jonathanrocher) [reply app_common/traitsui/dict_editors.py:37]: Agreed.

File: app_common/traitsui/dict_editors.py (hunk 2 above)
Comment (jonathanrocher) [reply app_common/traitsui/dict_editors.py:70]: Actually, never mind, that should be implemented in the downstream code since the buttons are lost when embedding this class in another view.

Comment (chamshop) [reply app_common/traitsui/dict_editors.py:79]: No need to implement it here. I would just consider removing the Cancel button.
